cmd/dra-example-kubeletplugin/driver.go

```diff
@@ -111,51 +85,34 @@ func (d *driver) NodePrepareResources(ctx context.Context, req *drapbv1.NodePrep
 }
 
 func (d *driver) nodePrepareResource(ctx context.Context, claim *drapbv1.Claim) *drapbv1.NodePrepareResourceResponse {
-	logger := klog.FromContext(ctx)
-	var err error
-	var prepared []string
-	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		prepared, err = d.prepare(ctx, claim.Uid)
-		if err != nil {
-			return fmt.Errorf("error allocating devices for claim '%v': %v", claim.Uid, err)
-		}
-
-		updatedSpec, err := d.state.GetUpdatedSpec(&d.nascrd.Spec)
-		if err != nil {
-			return fmt.Errorf("error getting updated CR spec: %v", err)
+	if len(claim.StructuredResourceHandle) == 0 {
+		return &drapbv1.NodePrepareResourceResponse{
+			Error: "driver only supports structured parameters",
 		}
+	}
 
-		err = d.nasclient.Update(ctx, updatedSpec)
-		if err != nil {
-			if err := d.state.Unprepare(claim.Uid); err != nil {
-				logger.Error(err, "Failed to unprepare after Update", "claim", claim.Uid)
-			}
-			return err
+	allocated, err := d.getAllocatedDevices(ctx, claim)
+	if err != nil {
+		return &drapbv1.NodePrepareResourceResponse{
+			Error: fmt.Sprintf("error allocating devices for claim %v: %v", claim.Uid, err),
 		}
+	}
 
-		return nil
-	})
-
+	prepared, err := d.state.Prepare(claim.Uid, allocated)
```

Comment:
This looks like claim preparation logic is implemented it `state`. I expected most of it here, in the driver.go

Reply:
What does that leave for `state` to do then? Are you suggesting I get rid of the `state` indirection altogether?

Reply:
It would leave at least checkpointing and probably other state-related operations. The actual device preparation should be decoupled from the state management in my opinion.

Reply:
I'm not sure how I feel about pulling the list of allocatable devices up into the driver -- this is the main thing that the state abstraction maintains, and "preparation" at this level is basically checking if the allocated device is actually something in the allocation list and then then checkpointing it. Maybe you just don't like the name `prepare()` for this lower-level operation? Do you have another suggestion.

Reply:
I had an impression that *all* resource preparation code is in the state.prepare. If this is not the case, than I'm ok with this.